File: rust/src/arena.rs
use std::{thread::sleep, time::Duration, io, ops::Range};
use rand::{thread_rng, Rng};

use crate::entity::*;

#[derive(PartialEq)]
enum States {
    Idle,
    TurnA,
    TurnB,
    Final
}

pub struct Arena {
    entity_a: Entity,
    entity_b: Entity,
    state: States
}

impl Arena {
    pub fn new(entity_a: Entity, entity_b: Entity) -> Arena {
        Arena {
            entity_a,
            entity_b,
            state: States::Idle
        }
    }

    pub fn fight(&mut self) {
        let randomize = thread_rng().gen_range::<i32, Range<i32>>(0..2);
        self.state = match randomize {
            1 => States::TurnA,
            _ => States::TurnB
        };
        let mut turn = 1;
        while self.state != States::Final {
            let a_damage = assign_attack_damage(50);
            let b_damage = assign_attack_damage(50);

            match self.state {
                States::TurnA => {
                    self.entity_a.deal_damage_to(&mut self.entity_b, a_damage);                       
                    self.display_health();
                    println!("============Turn {}===============", turn);
                    if self.entity_b.hp > 0 {
                        println!("{} Turn!", self.entity_b.name);
                    }
                    self.state = States::TurnB;
                },
                States::TurnB => {
                    self.entity_b.deal_damage_to(&mut self.entity_a, b_damage); 
                    self.display_health();
                    println!("============Turn {}===============", turn);
                    if self.entity_a.hp > 0 {
                        println!("{} Turn!", self.entity_a.name);
                    }
                    self.state = States::TurnA;
                },
                _ => {}
            }

            if self.entity_a.hp < 1 {
                println!("{} Wins!", self.entity_b.name);
                self.state = States::Final;
                continue;
            }

            if self.entity_b.hp < 1 {
                println!("{} Wins!", self.entity_a.name);
                self.state = States::Final;
                continue;
            }

            turn += 1;
            sleep(Duration::from_millis(1000));
        }
        self.prompt();
    }

    fn prompt(&mut self) {
        println!("===================================");
        println!("Type 0 or above to continue the game, type -1 to exit!");
        let mut input = String::new();
        if io::stdin().read_line(&mut input).is_ok() {
            let num = input.trim().parse::<i32>().unwrap_or(-1);
            if num > -1 {
                self.entity_a.hp = 100;
                self.entity_b.hp = 100;
                self.fight();
            }
        }
    }

    fn display_health(&self) {
        println!("{} has {} HP.", self.entity_a.name, self.entity_a.hp); 
        println!("{} has {} HP.", self.entity_b.name, self.entity_b.hp); 
    }
}

File: fsharp/Arena.fs
namespace TurnBasedFSharp
open System.Threading
open System

type States =
    | IdleState = 0
    | TurnAState = 1
    | TurnBState = 2
    | FinalState = 3

type Arena(contestA: Entity, contestB: Entity) =
    let random = new Random()
    let contestA = contestA
    let contestB = contestB
    let mutable states = States.IdleState

    member this.Fight() =
        let randomized = random.Next() % 2
        states <- 
            match randomized with 
                | 1 -> States.TurnAState 
                | _ -> States.TurnBState
        let mutable turn = 1
        while states <> States.FinalState do
            let (aDamage, bDamage) = (contestA.AssignAttackDamage (50, random), 
                                    contestB.AssignAttackDamage (50, random)) 
            Console.WriteLine("============Turn {0}===============", turn)

            match states with
                | States.TurnAState ->
                    contestA.DealDamageTo(contestB, aDamage)
                    this.DisplayHealth()

                    if contestB.Hp > 0 then
                        Console.WriteLine("{0} Turn!\n", contestB.Name)
                    states <- States.TurnBState
                | States.TurnBState ->
                    contestB.DealDamageTo(contestA, bDamage)
                    this.DisplayHealth()
                    if contestA.Hp > 0 then
                        Console.WriteLine("{0} Turn!\n", contestA.Name)
                    states <- States.TurnAState
                | _ -> printfn "Done!" 
            
            if contestA.Hp < 1 then
                states <- States.FinalState
                Console.WriteLine("{0} Wins!", contestB.Name)
            elif contestB.Hp < 1 then
                states <- States.FinalState
                Console.WriteLine("{0} Wins!", contestA.Name)
            else
                turn <- turn + 1
                Thread.Sleep(1000);
        this.Prompt()

                    
    
    member _.DisplayHealth(): unit =
        Console.WriteLine("{0} has {1} HP", contestA.Name, contestA.Hp)
        Console.WriteLine("{0} has {1} HP", contestB.Name, contestB.Hp)

    member this.Prompt() =
        printfn "Type 0 or above to continue the game, type -1 to exit!"
        let parsed = Console.ReadLine()
        let (success, result) = Int32.TryParse (parsed)
        if success then
            if result > -1 then
                contestA.Hp <- 100
                contestB.Hp <- 100
                this.Fight()
        else
            printfn "Not a valid number!, exiting the program"        
